src/hotstuff/protocol.rs

```diff
@@ -134,23 +560,158 @@ pub(crate) struct HotStuffConfiguration {
     pub(crate) keypair: Keypair,
 }
 
-/// Internal state of the [HotStuff] protocol. Stores the [Vote] and [NewView] messages collected
-/// in the current view. This state should always be updated on entering a view.
-struct HotStuffState {
-    vote_collector: VoteCollector,
-    new_view_collector: NewViewCollector,
+#[derive(Debug)]
+pub enum HotStuffError {
+    BlockTreeError(BlockTreeError)
 }
 
-impl HotStuffState {
-    /// Returns a fresh [HotStuffState] for a given view, leader, and validator set.
-    fn new(
-        config: HotStuffConfiguration,
-        view: ViewNumber,
-        validator_set: ValidatorSet,
-    ) -> Self {
-        Self {
-            vote_collector: VoteCollector::new(config.chain_id, view, validator_set.clone()),
-            new_view_collector: NewViewCollector::new(validator_set),
+impl From<BlockTreeError> for HotStuffError {
+    fn from(value: BlockTreeError) -> Self {
+        HotStuffError::BlockTreeError(value)
+    }
+}
+
+/// Captures the state of progress in a view. Keeping track of the [ViewStatus] is important for
+/// ensuring that a leader can only propose once in a view - any subsequent proposals will be
+/// ignored.
+/// 
+/// The [ViewStatus] can signal either of the above:
+/// 1. The view is in progress: proposals and nudges from a valid leader (proposer) can be accepted.
+/// 2. The leader with a given public key has already proposed or nudged: no more proposals or nudges
+///    from this leader can be accepted.
+/// 3. The view is completed (for use during the validator set update period): all leaders for the
+///    view have proposed or nudged, hence no more proposals or nudges can be accepted in this view.
+/// 
+/// Note that a variable of this type is stored in memory allocated to the program at runtime, rather
+/// than persistent storage. This is because losing the information stored in [ViewStatus], unlike
```

Comment:
Lots of links to `ViewStatus` in this itemdoc. If a definition has been linked to "recently", let's not link to it again (just like in Wikipedia).

Additionally, if something is being referred to in text which is a definition in the source code, let's format it like code, so "`ViewStatus`" instead of just "ViewStatus". I notice that, for example, the cargo docs for the Rust standard library is formatted in this way.